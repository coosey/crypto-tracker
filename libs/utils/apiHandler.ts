import { NextApiRequest, NextApiResponse } from 'next';

interface ApiHandlerConfig {
  url: string;
  method?: 'GET' | 'POST' | 'PUT' | 'DELETE';
  headers?: Record<string, string>;
  cacheConfig?: {
    enabled?: boolean;
    sMaxAge?: number;
    staleWhileRevalidate?: number;
  };
  queryDefaults?: Record<string, string | number>;
  requiredParams?: string[];
};

function urlQueryBuilder(query: Record<string, any>, urlBuilder: string) {
  if (Object.keys(query).length === 0) return urlBuilder;
  const queryParams = new URLSearchParams();
  for (const [key, value] of Object.entries(query)) {
    if (Array.isArray(value)) {
      value.forEach(v => queryParams.append(key, v));
    } else if (value) {
      queryParams.append(key, value as string);
    }
  }
  return `${urlBuilder}?${queryParams.toString()}`;
};

export function createApiHandler<T>(config: ApiHandlerConfig) {
  return async function handler(
    req: NextApiRequest,
    res: NextApiResponse<T | { error: string }>
  ) {
    try {
      if (config?.cacheConfig?.enabled !== false) {
        res.setHeader(
          'Cache-Control',
          `public, s-maxage=${config?.cacheConfig?.sMaxAge ?? 1800}, stale-while-revalidate=${config?.cacheConfig?.staleWhileRevalidate ?? 3600}`
        );
      }

      // Handle query parameters with defaults
      const query = {
        ...config?.queryDefaults,
        ...req?.query,
      };

      // Validate required parameters
      if (config?.requiredParams) {
        const missingParams = config?.requiredParams?.filter?.(
          param => !query?.[param]
        );
        if (missingParams?.length > 0) {
          return res.status(400).json({
            error: `Missing required parameters: ${missingParams?.join(', ')}`,
          });
        }
      }
      // Create a copy of the request query to avoid mutation
      let queryCopy = { ...req.query };

      // Build final URL
      let finalUrl = config?.url;
      // Replace path parameters if they exist in the URL i.e. /coins/{id}/market_chart
      if (config?.url?.includes?.('{')) {
        for (const [key, value] of Object.entries(queryCopy)) {
          if (config?.url?.includes?.(`{${key}}`)) {
            finalUrl = finalUrl?.replace?.(`{${key}}`, value as string);
            delete queryCopy[key];
          }
        }
      }
      // Then append query parameters
      if (Object.keys(queryCopy)?.length > 0) {
        finalUrl = urlQueryBuilder(queryCopy, finalUrl);
      }

      const response = await fetch(finalUrl, {
        method: config?.method || 'GET',
        headers: {
          'content-type': 'application/json',
          ...config?.headers,
        },
      });

      if (!response?.ok) {
        throw {
          status: response?.status,
          message: `API request failed with status ${response?.status}`,
        };
      }
      let data = await response.json();

      res.status(200).json(data);
    } catch (error) {
      console.error('API handler error:', error);
      res.status((error as any)?.status || 500).json({
        error: (error as Error)?.message || 'An unknown error occurred',
      });
    }
  };
}