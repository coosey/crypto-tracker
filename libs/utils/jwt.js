import jwt from 'jsonwebtoken';

export const generateToken = (payload, secret, expiresIn = '1h') => {
  return jwt.sign(payload, secret, { expiresIn });
};
