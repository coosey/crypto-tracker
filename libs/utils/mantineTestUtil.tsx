import { MantineProvider } from '@mantine/core';
import { ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';

const MantineTestProvider = ({ children }: { children: React.ReactNode }) => {
  return <MantineProvider>{children}</MantineProvider>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: MantineTestProvider, ...options });

export * from '@testing-library/react';
export { customRender as render };