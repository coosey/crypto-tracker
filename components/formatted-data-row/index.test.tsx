import '@testing-library/react';
import { render, screen } from 'libs/utils/mantineTestUtil';
import { FormattedDataRow } from '.';

describe('FormattedDataRow Component', () => {
  it('Should render FormattedDataRow Component correctly', () => {
    render(
      <FormattedDataRow
        rowName="Test"
        rowPrice={10000}
        priceDiff24High={100}
        priceDiff24Low={20}
        priceDiff7High={80}
        priceDiff7Low={40}
      />
    );
    const formattedDataRow = screen.getByText('Test');
    expect(formattedDataRow).toMatchSnapshot();
  });
  it('Should render HoverCard correctly', () => {
    render(
      <FormattedDataRow
        rowName="HoverCard"
        rowPrice={10000}
        hoverCard
        hoverCardName="HoverCard Test"
        hoverCardDescription="This is a test HoverCard description"
      />
    );
    const formattedDataRow = screen.getByText('HoverCard');
    expect(formattedDataRow).toMatchSnapshot();
  });
  it('Should render Market Cap correctly', () => {
    render(
      <FormattedDataRow
        rowName="Market Cap"
        rowPrice={10000}
        hoverCard
        hoverCardName="Market Cap = Current Price x Circulating Supply"
        hoverCardDescription="Refers to the total market value of a cryptocurrency’s circulating supply. It is similar to the stock market’s measurement of multiplying price per share by shares readily available in the market (not held & locked by insiders"
      />
    );
    const formattedDataRow = screen.getByText('Market Cap');
    expect(formattedDataRow).toMatchSnapshot();
  });
  it('Should render Volume correctly', () => {
    render(
      <FormattedDataRow
        rowName="Volume"
        rowPrice={10000}
        hoverCard
        hoverCardName="Volume"
        hoverCardDescription="A measure of how much of a cryptocurrency was traded in the last 24 hours."
      />
    );
    const formattedDataRow = screen.getByText('Volume');
    expect(formattedDataRow).toMatchSnapshot();
  });
  it('Should render FDV correctly', () => {
    render(
      <FormattedDataRow
        rowName="FDV"
        rowPrice={10000}
        hoverCard
        hoverCardName="Fully Diluted Valuation (FDV) = Current Price x Total Supply"
        hoverCardDescription="Fully Diluted Valuation (FDV) is the theoretical market capitalization of a coin if the entirety of its supply is in circulation, based on its current market price. The FDV value is theoretical as increasing the circulating supply of a coin may impact its market price. Also depending on the tokenomics, emission schedule or lock-up period of a coin's supply, it may take a significant time before its entire supply is released into circulation."
      />
    );
    const formattedDataRow = screen.getByText('FDV');
    expect(formattedDataRow).toMatchSnapshot();
  });
  it('Should render Circulating Supply correctly', () => {
    render(
      <FormattedDataRow
        rowName="Circulating Supply"
        rowPrice={10000}
        hoverCard
        hoverCardName="Circulating Supply"
        hoverCardDescription="The amount of coins that are circulating in the market and are tradeable by the public. It is comparable to looking at shares readily available in the market (not held & locked by insiders, governments)."
      />
    );
    const formattedDataRow = screen.getByText('Circulating Supply');
    expect(formattedDataRow).toMatchSnapshot();
  });
  it('Should render Total Supply correctly', () => {
    render(
      <FormattedDataRow
        rowName="Total Supply"
        rowPrice={10000}
        hoverCard
        hoverCardName="Total Supply = Onchain supply - burned tokens"
        hoverCardDescription="The amount of coins that have already been created, minus any coins that have been burned (removed from circulation). It is comparable to outstanding shares in the stock market."
      />
    );
    const formattedDataRow = screen.getByText('Total Supply');
    expect(formattedDataRow).toMatchSnapshot();
  });
  it('Should render Max Supply correctly', () => {
    render(
      <FormattedDataRow
        rowName="Max Supply"
        rowPrice={10000}
        hoverCard
        hoverCardName="Max Supply = Theoretical maximum as coded"
        hoverCardDescription="The maximum number of coins coded to exist in the lifetime of the cryptocurrency. It is comparable to the maximum number of issuable shares in the stock market."
      />
    );
    const formattedDataRow = screen.getByText('Max Supply');
    expect(formattedDataRow).toMatchSnapshot();
  });
});
