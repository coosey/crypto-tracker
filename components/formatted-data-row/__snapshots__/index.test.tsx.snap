// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormattedDataRow Should render Circulating Supply correctly 1`] = `
<span
  class="table-row_name"
>
  Circulating Supply
</span>
`;

exports[`FormattedDataRow Should render FDV correctly 1`] = `
<span
  class="table-row_name"
>
  FDV
</span>
`;

exports[`FormattedDataRow Should render HoverCard correctly 1`] = `
<span
  class="table-row_name"
>
  HoverCard
</span>
`;

exports[`FormattedDataRow Should render Market Cap correctly 1`] = `
<span
  class="table-row_name"
>
  Market Cap
</span>
`;

exports[`FormattedDataRow Should render Max Supply correctly 1`] = `
<span
  class="table-row_name"
>
  Max Supply
</span>
`;

exports[`FormattedDataRow Should render Total Supply correctly 1`] = `
<span
  class="table-row_name"
>
  Total Supply
</span>
`;

exports[`FormattedDataRow Should render Volume correctly 1`] = `
<span
  class="table-row_name"
>
  Volume
</span>
`;
